---
title: "GSE96870 for the intro lesson"
output: html_document
date: "`r Sys.Date()`"
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages

```{r}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(tidyverse)
  library(org.Mm.eg.db)
  library(biomaRt)
})
```

# Read and annotate the data

```{r}
se <- readRDS("GSE96870_se.rds")
## subset according to the differential expression vignette
se <- se[, se$tissue == "Cerebellum"]

## Add gene description to rowData
ensembl <- useDataset("mmusculus_gene_ensembl", useMart("ensembl"))

att1  <- c("external_gene_name",
           "ensembl_gene_id",
           "external_synonym",
           "chromosome_name",
           "gene_biotype",
           "phenotype_description")

description1 <- getBM(attributes = att1,
                      mart = ensembl,
                      filters = "external_gene_name",
                      values = rownames(se))

att2  <- c("external_gene_name",
           "hsapiens_homolog_associated_gene_name")

description2 <- getBM(attributes = att2,
                      mart = ensembl,
                      filters = "external_gene_name",
                      values = rownames(se))

description <- description1 %>%
  filter(!duplicated(external_gene_name)) %>%
  left_join(description2 %>%
              filter(!duplicated(external_gene_name)))

description[description == ""] <- NA

rownames(description) <- description$external_gene_name

keep <- rownames(se)[rownames(se) %in% rownames(description)]
se <- se[keep, ]

rowData(se) <- cbind(rowData(se), description[keep, ])
se
rowData(se)
```

# Select a subset of genes

```{r}
resTime <- data.frame(readRDS("resTime.rds")) %>%
  rownames_to_column("gene") %>%
  as_tibble() %>%
  filter(gene %in% keep) %>%
  na.omit() %>%
  filter(!grepl("^LOC", gene)) %>%
  filter(!grepl("Rik$", gene))

resSex <- data.frame(readRDS("resSex.rds")) %>%
  rownames_to_column("gene") %>%
  as_tibble() %>%
  filter(gene %in% keep) %>%
  na.omit() %>%
  filter(!grepl("^LOC", gene)) %>%
  filter(!grepl("Rik$", gene))
```


```{r}
sel_genes <- c(
  ## 500 top time DE
  resTime %>%
    arrange(padj) %>%
    head(500) %>%
    pull("gene"),
  ## 500 top sex DE
  resSex %>%
    arrange(padj) %>%
    head(500) %>%
    pull("gene"),
  ## 500 non-DE time
  resTime %>%
    filter(padj > 0.2) %>%
    head(500) %>%
    pull("gene")
)
sel_genes <- unique(sel_genes)
length(sel_genes)
```

# Prepare data for the intro lesson


```{r}
se2 <- se[sel_genes, ]

## TODO
## - add product (contains NA)
## - add p-values

## NEW DATA
## - add other variables such as function/location (see biomart) that
##   have NAs
## - chromosome

x <- MultiAssayExperiment::longFormat(se2) %>%
  as_tibble() %>%
  dplyr::rename("gene" = "rowname",
                "sample" = "colname",
                "expression" = "value")

cd <- colData(se2) %>%
  as_tibble() %>%
  mutate(age = as.numeric(sub(" weeks", "", age))) %>%
  mutate(time = as.numeric(sub("Day", "", time))) %>%
  mutate(infection = sub("[ -]", "", infection)) %>%
  dplyr::select(-title)

rd <- rowData(se2) %>%
  as_tibble() %>%
  dplyr::select(-external_gene_name) %>%
  dplyr::select(-gbkey)

```

```{r}
rnaseq <- full_join(x, cd,
                    by = c("sample" = "geo_accession"))
rnaseq <- full_join(rnaseq, rd)
```

# Serialise data

```{r}
saveRDS(se2, file = "se2.rds")
saveRDS(rnaseq, file = "rnaseq.rds")

## writing with quoting to handle cases with a ',' in their product
## such as "family with sequence similarity 193, member B" for Fam193b
## and that break import.
write_csv(rnaseq, "rnaseq.csv", quote = "all")
```

# Session info

```{r}
sessionInfo()
```
